XiEditor/StyleMap.swift

```diff
@@ -61,15 +57,46 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
 class StyleMap {
     private var map: [Style?] = []
 
-    func defStyle(json: [String: AnyObject]) {
+    func defStyle(for fromFont: NSFont, json: [String: AnyObject]) {
         guard let styleID = json["id"] as? Int else { return }
-        let fgColor = json["fg_color"] as? UInt32 ?? 0xFF000000
-        let bgColor = json["bg_color"] as? UInt32 ?? 0
-        let weight = json["weight"] as? UInt16 ?? 400
-        let underline = json["underline"] as? Bool ?? false
+
+        var attrs: [String: Any] = [
+            NSForegroundColorAttributeName: json["fg_color"] as? NSColor ?? NSColor.black,
+        ]
+
+        if let bgColor = json["bg_color"] as? NSColor, bgColor.alphaComponent != 0.0 {
+            attrs[NSBackgroundColorAttributeName] = bgColor
+        }
+
+        if let underline = json["underline"] as? Bool, underline {
+            attrs[NSUnderlineStyleAttributeName] = NSUnderlineStyle.styleSingle.rawValue
+        }
+
+        let fm = NSFontManager.shared()
+        var font: NSFont?
+
         let italic = json["italic"] as? Bool ?? false
+        // convert to NSFont weight: (100-500 -> 2-6 (5 normal weight), 600-800 -> 8-10, 900 -> 12
+        let w = Float(json["weight"] as? Int ?? fm.weight(of: fromFont))
+        let weight = Int(floor(1 + w * (0.01 + 3e-6 * w)))
```

Comment:
Yes, that would be less code and I guess the comment above is enough to not require a reader to quickly understand the formula. One (not clearly mentioned) part of my question was, if we want to add this case at all, since it is somehow unclear whether `1000` is already a valid CSS font weight. :)